### PR TITLE
bug in checking salt length is fixed

### PR DESCRIPTION
**Pull request by srmarjani**
I committed by --no-verify flag. When pre-commit hook  tried to run npm test there was errors.
But when I run it by myself all of test passes.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -6,25 +6,23 @@ const kObj = Symbol('object')
...
-    // static salt to be used for key derivation, not great for security,
-    // but better than nothing
-    var salt = 'mq9hDxBVDbspDR6nLfFT1g=='
+    salt = Buffer.alloc(sodium.crypto_pwhash_SALTBYTES)
 
     if (options.salt) {
-      salt = options.salt
+      salt = Buffer.from(options.salt, 'ascii')
+    } else {
+      sodium.randombytes_buf(salt)
```
Comment (mcollina) [inline index.js]: Unfortunately this is not correct. Multiple instances of the same app must share the same salt.

**Comment by srmarjani** [comment]
Hi
I fixed the problem. 
Plz recheck it.

**Comment by srmarjani** [comment]
> Can you add unit tests to cover the new functionality?

I though that there is no need for unit test. Because I have not added new feature.

**Comment by mcollina** [comment]
You are fixing a bug, so a unit test is in order.

**Comment by srmarjani** [comment]
> You are fixing a bug, so a unit test is in order.

OK. I will add unit test. :))
But Are you ok with changes that I made to code?
Any suggestion?

File: index.js [hunk 2]
```diff
@@ -6,25 +6,23 @@ const kObj = Symbol('object')
...
-    // static salt to be used for key derivation, not great for security,
-    // but better than nothing
-    var salt = 'mq9hDxBVDbspDR6nLfFT1g=='
+    salt = Buffer.alloc(sodium.crypto_pwhash_SALTBYTES)
 
     if (options.salt) {
-      salt = options.salt
+      salt = Buffer.from(options.salt, 'ascii')
+    } else {
+      salt = Buffer.from('mq9hDxBVDbspDR6n', 'ascii')
```
Comment (mcollina) [inline index.js]: I would not change the actual value of the default, can you restore that?

Also, kee comment.

File: index.js [hunk 3]
```diff
@@ -15,16 +15,16 @@ module.exports = fp(function (fastify, options, next) {
 
     // static salt to be used for key derivation, not great for security,
     // but better than nothing
-    var salt = 'mq9hDxBVDbspDR6nLfFT1g=='
+    var salt = 'mq9hDxBVDbspDR6n'
```
Comment (mcollina) [inline index.js]: I would prefer the default salt __not__ be changed. In other terms store it as a buffer.

File: index.js [hunk 4]
```diff
@@ -15,16 +15,16 @@ module.exports = fp(function (fastify, options, next) {
...
 
     if (options.salt) {
       salt = options.salt
     }
 
-    salt = Buffer.from(salt, 'base64')
+    salt = Buffer.from(salt, 'ascii')
```
Comment (mcollina) [inline index.js]: We should likely not do this is if `salt` is a buffer.

File: index.js [hunk 5]
```diff
@@ -15,16 +15,16 @@ module.exports = fp(function (fastify, options, next) {
...
 
     if (Buffer.byteLength(salt) !== sodium.crypto_pwhash_SALTBYTES) {
-      return next(new Error('salt must be length 24'))
+      return next(new Error('salt must be length 16'))
```
Comment (mcollina) [inline index.js]: I would use the value from `sodium.crypto_pwhash_SALTBYTES` to avoid future issues.

Comment (srmarjani) [reply index.js]: I've used `sodium.crypto_pwhash_SALTBYTES` but after refactoring I forget it. I will fix it.

File: index.js (hunk 3 above)
Comment (srmarjani) [reply index.js]: The previous salt was 24 character.When It converted  to Buffer by `Buffer.from(salt, 'base64')` the length of result buffer is 16 Byte. So I changed the length of default salt to 16 character  and for converting  it to buffer I've used `Buffer.from(salt, 'ascii')`

File: index.js (hunk 4 above)
Comment (srmarjani) [reply index.js]: We can check salt, If it is a buffer there is no need to `Buffer.from()`

File: index.js (hunk 3 above)
Comment (mcollina) [reply index.js]: You are altering the actual bytes in the Buffer resulting the parsed operation. Truncating the string is not enough.

Comment (srmarjani) [reply index.js]: I thought that this string is a random string. How can I generate it.

Comment (mcollina) [reply index.js]: It was a random string. However it's our current default, and changing it is an unneeded semver-major change.

You can do: 

```js
const defaultSalt = Buffer.from('mq9hDxBVDbspDR6nLfFT1g==', 'base64')
```

and then pass it down as a buffer.

File: index.js (hunk 4 above)
Comment (mcollina) [reply index.js]: Exactly.

File: test/secret.js [hunk 6]
```diff
@@ -1,48 +1,161 @@
...
+      t.deepEqual(JSON.parse(response.payload), {
+        some: 'data'
+      })
+    })
+  })
 })
 
-t.tearDown(fastify.close.bind(fastify))
-t.plan(5)
+tap.test('using secret with salt.Salt is string', function (t) {
```
Comment (mcollina) [inline test/secret.js]:
```suggestion
tap.test('using secret with salt as string', function (t) {
```

File: test/secret.js [hunk 7]
```diff
@@ -1,48 +1,161 @@
...
-  method: 'POST',
-  url: '/',
-  payload: {
-    some: 'data'
-  }
-}, (error, response) => {
-  t.error(error)
-  t.equal(response.statusCode, 200)
-  t.ok(response.headers['set-cookie'])
+tap.test('using secret with salt.Salt is baffer', function (t) {
```
Comment (mcollina) [inline test/secret.js]:
```suggestion
tap.test('using secret with salt as buffer', function (t) {
```

**Comment by mcollina** [comment]
Can you apply the suggested changes so I can land this?

**Pull request merged**